### PR TITLE
Add PufferFishStateChangeEvent

### PR DESCRIPTION
**Pull request by HexedHero**
Added a new Pufferfish event for when it inflates or deflates (Changes puff state), was surprised there wasn't an event for this yet and thought it could be cool in some niche plugins such as changing all pufferfishes to be fully inflated and never let them deflated or when it inflates shoot arrows out.

Tested with:
```JAVA
    private boolean cancel = false;
    
    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    public void a(PufferFishStateChangeEvent event) {
        handleEvent(event, event.getEntity());
    }
    
    private void handleEvent(Cancellable event, Entity entity) {
        Bukkit.broadcastMessage("Fired " + event.toString());
        entity.getWorld().spawnParticle(Particle.COMPOSTER, entity.getLocation(), 4, 0.05, 0.05, 0.05, 0);
        if (cancel) { 
            event.setCancelled(true);
        }
    }
    
    // Say "TOGGLE" to toggle cancelling the event
    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    public void b(PlayerChatEvent event) {
        if (event.getMessage().equalsIgnoreCase("TOGGLE")) {
            cancel = !cancel;
            event.getPlayer().sendMessage("Toggled to " + cancel);
            event.setCancelled(true);
        }
    }
```

### CODE REVIEW

**Comment by kezz** [comment]
Could each event extend a common `PufferfishChangeEvent` or something like that? Love the "puff state" name btw :')

Or alternatively just one event that has a getter to check if it's inflating or deflating.

**Comment by HexedHero** [comment]
> Could each event extend a common `PufferfishChangeEvent` or something like that? Love the "puff state" name btw :')
> 
> Or alternatively just one event that has a getter to check if it's inflating or deflating.

I think I'd like to advice on what to do here, should I make a common event that they both extend or just make it a single event with a getter? If so what would the event name be, PufferFishChangeEvent, PufferFishPuffEvent, PufferFishFlateEvent? I'm not so sure what is the best thing to do here. I'm leaning towards just a single event but the naming is hard.

File: Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
++
++/**
++ * Called just before a {@link PufferFish} inflates or deflates.
++ */
++public class PufferFishFlateEvent extends EntityEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
++    private boolean cancelled;
++    private int newPuffState;
++
++    public PufferFishFlateEvent(@NotNull PufferFish entity, int newPuffState) {
```
Comment (Trigary) [inline Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch]: You forgot to assign `newPuffState`

File: Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
++
++    @NotNull
++    @Override
++    public PufferFish getEntity() {
++        return (PufferFish) entity;
++    }
++
++    /**
++     * Get the <strong>new</strong> puff state of the {@link PufferFish}.
++     * <br><br>
```
Comment (Trigary) [inline Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch]: I believe Paper prefers `<p>` blocks over `<br><br>`: I was asked to change something like this once.

**Review by Trigary — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
++import org.bukkit.entity.PufferFish;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.entity.EntityEvent;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called just before a {@link PufferFish} inflates or deflates.
++ */
++public class PufferFishFlateEvent extends EntityEvent implements Cancellable {
```
Comment (Proximyst) [inline Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch]: Unnatural name; perhaps `PufferFishStateChangeEvent`, `PufferFishInflationEvent`?

**Review by Proximyst — APPROVED** [review]
looks fine

File: Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch (hunk 3 above)
Comment (Geolykt) [reply Spigot-API-Patches/0296-Add-PufferFishFlateEvent.patch]: I don't think that PufferFishInflationEvent is a good one given that it also handles deflation

**Review by electronicboy — COMMENTED** [review]
is State really the best term for this? otherwise, LGTM

**Comment by electronicboy** [comment]
I guess that's what vanilla calls it, bleh

**Review by kennytv — APPROVED** [review]
Thank you, "reased", and you made this guy very happy.
javaw_2019-11-27_00-22-17

**Pull request merged**